UI Docs update (#269)
space-wizards/docs: src/en/ss14-by-example/ui-and-you.md

`Panel`, and `Label`.

The rest of the style classes are defined for a specific generic UI element.
Some common style classes are as follows:

-   `OpenLeft`: Makes the button flat on its left side
-   `OpenRight`: Makes the button flat on its right side
-   `OpenBoth`: Makes the button flat on both sides; square
-   `LabelSubtext`: Makes the label smaller and a more muted color
-   `LabelKeyText`: Makes the label bold and a highlight color
-   `LabelWeak`: Weak is the opposite of strong; makes the label a more muted color

There are many more than this, but if you want to know exactly what a given label
does, its as simple as looking at the field's usages, and reading the style rule definition.
This is where it may be helpful to use the Rider IDE, as it has a fantastic implementation of this feature,
but its probably possible in other IDEs as well.

```admonish tip
In general, if you are doing UI dev, I would recommend using the Rider IDE. It
eats up quite a bit of RAM, but it provides autocomplete in XAML files, a lot
of really nice auto-refactoring and searching features, and very decent git
integration. Give it a shot!

> ```

Author: I'm pretty sure rider is better? I'll just make a note about VSCode in case im dumb

## Writing Styles

This section concerns style rules. For most UIs, editing these will be unnecessary,
however you should ALWAYS prefer to use style classes instead of hardcoding colors
or resources that could be commonly used.

### All hail the mighty `Sheetlet`

It's important to understand that basically, a stylesheet is a massive list of
every single style rule. Instead of manually
making a giant list of style rules (because that would be ridiculous... haha... ha....),
the responsibility of chipping into this list is distributed between many Sheetlets.
Each Sheetlet returns a small chunk of style rules, which is agglomerated into the

> final list at the end.

Author: I'm going to be so real I'm not sure. I just elaborated a bit on the specificity of style rules a bit. That's the only surefire way to get style rules to behave.

> final list at the end.

Author: Also I'm leaking the details so people fully understand why stylenano existed

```admonish note
Previously every single style rule was in one giant list: `StyleNano.cs`, a 1600
line pit of despair where dreams went to die.
```

There are, primarily, two types of Sheetlets:

-   **Generic Sheetlets**: These go in `Content.Client/Stylesheets/Sheetlets`.
    These stylesheets concern generic UI elements used in many different UIs, and

>     should be written generically to work with any stylesheet.

Reviewer: you mean 'should be written to work with any OTHER stylesheet'?

    with.

We will go into more detail about the specific conventions to follow for both of
these later.

All sheetlets should have the `[CommonSheetlet]` attribute. This is used so
stylesheets can find all the sheetlets that have this attribute and add their
styles to their list.

```admonish tip
Do not forget the `[CommonSheetlet]` attribute.
```

### Style Rules

Each style rule
has a few different components to it (This will look familiar to anyone who knows CSS):

**The selector**: This limits the elements that a style rule can affect. This is
made up of a few different parts:

-   `Type`: The type of element this rule affects. Anything inheriting from this
    type will be affected by this rule. This is similar to the element selector
    in css.
-   `StyleClasses`: The classes that the element must have to be affected by this
    rule. The element must have all of these classes to be affected by this rule.
-   `StyleIdentifier`: The identifier of the element. This is a unique identifier
    that can be used to target a specific element. This should be used sparingly,
    when there is only one instance of the element that needs to be styled in a

>     highly specific manner. There may only be one of these specified.

Author: I tried my best

        // ...and set its font color to the text color from the negative palette
        .FontColor(sheet.NegativePalette.Text),

    // select any `Label`...
    E<Label>()
        // ...with the class "LabelHeading"
        .Class(StyleClass.LabelHeading)
        // ...and set its font to a bold 16pt font
        .Font(sheet.BaseFont.GetFont(16, FontKind.Bold))
        // ...and its font color to the text color from the highlight palette
        .FontColor(sheet.HighlightPalette.Text)

    // select any `ContainerButton`...
     E<ContainerButton>()
        // ...with the class "button"
        .Class(ContainerButton.StyleClassButton)
        // ...and the class "ButtonSmall"
        .Class(StyleClass.ButtonSmall)
        // ...that is the parent of a `Label`,
        .ParentOf(E<Label>())
        // ...and set that `Label`'s font to an 8pt font
        .Font(sheet.BaseFont.GetFont(8))
];
```

### Death to Hardcoding!

You may have noticed before that there wasn't much hardcoding in the rule definitions.
This is because most style rules are used in multiple stylesheets, and they have
to be generic between them. There are a couple different utilities used to reduce

> hardcoding.

Reviewer 2: This paragraph feels a bit awkward to me. I think I'd reword this focusing more on the purpose behind avoiding hardcoding (increasing reusability of common patterns).

#### `ColorPalette`

There is actually a pretty robust (haha) color palette system to hopefully make
hardcoding colors unnecessary. I created a

> [codepen](https://codepen.io/aspiringLich/pen/VwOXdjd?editors=1000)

Reviewer 2: Linking to an external website like this feels like a recipe for a future dead or outdated link.  I'm not sure what the best alternative is, though.

Author: Replaced with an image.

to help visualize the colors. The stylesheets then use these palettes to set the following

> common palettes for sheetlets to reference:

Reviewer 2: This last sentence is weird. The stylesheets use palettes to set palettes? What does it means for sheetlets to reference something? I think it should be more direct about what actual effect palettes have.

-   `PrimaryPalette`: Used for foreground elements

> -   `SecondaryPalette`: Used for Background elements

Reviewer 2: ```suggestion
-   `SecondaryPalette`: Used for background elements
```

-   `PositivePalette`: A traditionally green palette used to represent success / good / full
-   `NegativePalette`: A traditionally red palette used to represent errors / bad / empty
-   `HighlightPalette`: Used to highlight headings or important elements

> In C#, you access the colors by accessing the

Reviewer: This palette section sadly seems a bit incoherent and scuffed - maybe best way is to show some examples?

properties on the `ColorPalette` class. From brightest to darkest, the properties
are (as of writing) arranged as so:

-   `+0`: `Text` `Base`
-   `-1`: `TextDark`, `Element`
-   `-2`: `BackgroundLight`, `PressedElement`
-   `-3`: `Background`

> -   `-4`: `BackgroundDark`, `DisabledElement`

Reviewer 2: What are the numbers here?

Although it would be simpler to have the palettes be arrays of colors, I noticed
that colors were used in basically three ways: Text, foreground elements, and
background elements. So I represented that in the palette itself! It just makes
the code a bit more readable.

All palettes are defined in `Palettes.cs` as static properties. You can read

> them from anywhere! Neat!

Reviewer 2: These last two paragraphs could be re-written. I think it'd be better to just describe the why of why this approach was taken.

#### `ISheetletConfig`

`ISheetletConfig` is intended to cut down on repeated code by providing shared
functionality and definitions between the stylesheets.
Any `Sheetlet` that requires the values in some instance of
`ISheetletConfig` should have a generic type constraint that requires the

> `ISheetletConfig` interface.

Reviewer 2: This needs a better example. What kind of repeated code? What functionality is being shared?

```cs
[CommonSheetlet] // don't forget `[CommonSheetlet]`!
public sealed class ExampleSheetlet<T> : Sheetlet<T> where T : PalettedStylesheet, IExampleConfig
```

`ISheetletConfig` also serves as a dependency check. When the stylesheets pull
all the sheetlets that have `[CommonSheetlet]`, they will first check that they
satisfy the type constraint before adding the rules to the stylesheet.

```admonish info
A sheetlet failing to satisfy the type constraints of any stylesheet will cause
the game to log an error. If your styles are not showing up, this may be why.
```

#### Resource Access

> You should access resources differently in sheetlets. Each stylesheet provides a
> list of resource roots to look in, when requesting a resource of a certain type.
> For example, the root for `NanotrasenStylesheet` for `TextureResource` is
> `/Textures/Interface/Nano`. Any texture requested with `GetTexture` will will
> append the provided relative path to the root, and return the texture if it exists.
> 

Reviewer 2: ```suggestion
Resources are accessed different in sheetlets than other parts of the codebase.
Each stylesheet provides a list of directories (roots) to use when a resource
is requested. For example, the `TextureResource` root in `NanotrasenStylesheet`
is `/Textures/Interface/Nano`. This means that any texture requested with
`GetTexture` will be searched for relative to this directory.
```

This system is built to work with any resource type, though currently only textures

> are used in the sheetlets.

Reviewer 2: This paragraph isn't really necessary. The section is already about accessing resources in general.

### Generic Sheetlets

> As stated earlier, generic sheetlets are used for generic UI elements that are
> used in many different UIs.
> 

Reviewer 2: ```suggestion
Generic sheetlets are used for common UI elements that are used throughout many
different UIs.
```

> -   You should always select elements with `.Class` and not `.Identifier`.

Reviewer 2: These bullet points need an introduction (here are some best practicies blah blah).

-   When accessing resources, use the `GetTextureOr` method to get the texture and
    provide a fallback root to use if the texture is not found within the stylesheets roots.
-   Avoid manual hardcoding of classes. When referencing classes,

>     please only use classes defined on the element being styled (in `StyleClass*` properties)

Reviewer 2: Using 'please' in technical documentation is very weird to me. In my opinion it should pretty much always be a 'you should' instead.

    or define your own in `StyleClass.cs`.

> -   Avoid manual hardcoding of colors. Use the palettes provided by the stylesheet
>     to set the colors of elements. There are very few cases where you will actually
>     need to manually hardcode a color in a generic sheetlet.

Reviewer 2: ```suggestion
-   Use the palettes provided by the set stylesheet instead of specifying 
    colors by hand. There are very few cases where you will actually need
    to manually hardcode a color in a generic sheetlet.
```

> -   If you need to access a resource that is not provided already, please add the path

Reviewer 2: Ditto the note on 'please' (and also anywhere else); I see one other usage earlier that I didn't mention too.

{
    public override StyleRule[] GetRules(T sheet, object config)
    {
        // cast the sheet into any of its required dependencies here
        ICheckboxConfig checkboxCfg = sheet;

        // get any textures / construct any complicated resources here
        var uncheckedTex = sheet.GetTextureOr(checkboxCfg.CheckboxUncheckedPath, NanotrasenStylesheet.TextureRoot);
        var checkedTex = sheet.GetTextureOr(checkboxCfg.CheckboxCheckedPath, NanotrasenStylesheet.TextureRoot);

        // and finally, define all the style rules and return a big 'ol list of them
        return
        [
            E<TextureRect>()
                .Class(CheckBox.StyleClassCheckBox)
                .Prop(TextureRect.StylePropertyTexture, uncheckedTex),
            E<TextureRect>()
                .Class(CheckBox.StyleClassCheckBox)
                .Class(CheckBox.StyleClassCheckBoxChecked)
                .Prop(TextureRect.StylePropertyTexture, checkedTex),
            E<BoxContainer>()
                .Class(CheckBox.StyleClassCheckBox)
                .Prop(BoxContainer.StylePropertySeparation, 10),
        ];
    }
}
```

</details>

### Specific Sheetlets

> As stated earlier, specific sheetlets are used for UI elements that are specific
> to a single UI. These sheetlets are located with the `*.xaml` file they are
> associated with.
> 

Reviewer 2: ```suggestion
Specific sheetlets are used in conjunction with UI elements that are used only
a few times, most often all in a single UI. These sheetlets are located in the
same directory as the `*.xaml` file they are associated with.
```

> -   You should prefer to select elements with `.Identifier` and not `.Class`.

Reviewer 2: This bullet list needs an intro stating what these points are about.

    `ISheetletConfig` you can access it through `ResCache` as normal.
-   You don't need to do the generic type constraints thing since the sheetlet
    should be specific to a single UI, and thus a single stylesheet.

<details>
<summary>Example Code (click to expand)</summary>

```cs
using Content.Client.Resources;
using Content.Client.Stylesheets.Redux;
using Content.Client.Stylesheets.Redux.SheetletConfigs;
using Content.Client.Stylesheets.Redux.Stylesheets;
using Robust.Client.Graphics;
using Robust.Client.UserInterface;
using Robust.Client.UserInterface.Controls;
// you need to add this line manually to access the helper methods
using static Content.Client.Stylesheets.Redux.StylesheetHelpers;

namespace Content.Client.Paper.UI;

// MAKE SURE TO INCLUDE THE [CommonSheetlet] ATTRIBUTE
[CommonSheetlet]
// which stylesheet is this sheetlet for
public sealed class PaperSheetlet : Sheetlet<NanotrasenStylesheet>
{
    public override StyleRule[] GetRules(NanotrasenStylesheet sheet, object config)
    {
        // define any IConfigs you need here
        IWindowConfig windowCfg = sheet;

        // get any textures / construct any complicated resources here

>         var paperBackground = ResCache.GetTexture("/Textures/Interface/Paper/paper_background_default.svg.96dpi.png")

Reviewer 2: This filename should be a static readonly respath. I don't think we validate those, but we probably will some day.

Author: Uh oh I don't think I do that in the like... actual PR. I'll... leave it for now.

            .IntoPatch(StyleBox.Margin.All, 16);
        var paperBox = new StyleBoxTexture
            { Texture = sheet.GetTexture(windowCfg.TransparentWindowBackgroundBorderedPath) };
        paperBox.SetPatchMargin(StyleBox.Margin.All, 2);

        // and finally, define all the style rules and return a big 'ol list of them
        return
        [
            E<PanelContainer>().Identifier("PaperContainer").Panel(paperBox),
            E<PanelContainer>()
                .Identifier("PaperDefaultBorder")
                .Prop(PanelContainer.StylePropertyPanel, paperBackground),
        ];
    }
}
```

</details>

### Making your own `Stylesheet`

> Creating a new stylesheet is pretty simple, so I won't go into it too much.

Reviewer 2: It's simple because you wrote the whole dang thing :P

Something like making a whole stylesheet is absolutely not simple to anyone else. Since this section doesn't really say anything at all about how to make a stylesheet, I think it's fine to just remove it, or mark that the section needs expanding (which is super fine to do!).

Also evil first person in technical documentation!!

Reviewer: On top of that i have to remark that it is critical to mention when to use stylesheet and when - not. Having several examples will be critical. This stuff is VERY not trivial for newcomers, and even for seasoned veterans - stylesheets does not always sound very alluring, compared to just setting couple of values in XAML or in control c-tor. And sometimes it can work.

One thing to keep in mind is how colors are chosen.

The colors are generated using the [OKLAB color space](https://bottosson.github.io/posts/oklab/),
which is better because something something human eyes something. When you choose
new colors for your stylesheet, it may be helpful to use an [OKLCH Color Picker](https://oklch.com)

> and modify an existing color.

Reviewer 2: Pedantry time: colors aren't really generated using a color space—in fact, a color on its own doesn't "have" a color space. A color space is really just a way of organizing colors (usually implicitly paired with a color model which is how we represent those colors—that's the difference between Oklab and OKLCH).

It's also fine to just hand off the explanation of why Oklab is cool to another website, but the gist of it is that it lets us interpolate between certain parameters in a way that visually preserves the unchanged variables. A transition between two colors of equal lightness will have perceptually uniform lightness along the entire transition. This ends up being very helpful for anything that ends up being 'what people see', as opposed to other color spaces and models that are more useful for (some) computations.

## Writing C# for UI

> > **TODO:** I don't feel confident enough in my knowledge to describe in detail

Reviewer 2: I think talking about actual UI implementation can be done in a different page, and you could definitely get away with not having to write about it in this PR if you want :P

> what to do and what not to do. This is just a general overview for now and
> should be updated.

The best way to learn how to write UI code is to look at existing code. Some UIs
definitely do some terrible things you should never replicate, but there are mountains
of terrible code in SS14, so this is not abnormal. I cannot teach familiarity with
the internals of this game, but I can give a  general overview.

Code you could reference:
-    Robotics Console
-    Cryo Pod

> -    Reagent Dispenser

Reviewer 2: BatteryMenu is probably the best example right now, expect for the fact that it is still BUI state based.

> {
>     if (!_uiSystem.TryGetOpenUi<JukeboxBoundUserInterface>(ent.Owner, JukeboxUiKey.Key, out var bui))
>         return;
>
>     bui.Reload();
> }
> ```
>
> Option 2. Have the BUI control be a dummy until the state comes in
>
> For UIs:
> Call TryOpenUi in shared where possible and the client should just handle it. Calling it from server will also still work similar to the old behavior.
>
> For messages:
> Use SendPredictedMessage where possible in BUIs. At some point this will likely become the default over SendMessage.
>
> Overall:
> Prefer to use the overloads that take in EntityUids instead of ICommonSession, this will make it easier to code NPCs who can interact with UIs in future.
>
> >
>
> -   There's a helper under this.CreateWindow<TWindow>() for BUIs that handles disposing + opening + close subscription for you.
> -   There's a method OnProtoReload that gets called on BUIs so you can override and handle it without having to manually subscribe on another system.
> -   I have added prototype reload support to some stuff.
> -   I have cleaned up a lot of BUI code. Windows now just raise events and the BUI itself handles message sending.
>

> > Some notes for future:

Author: ditto
